src/connection/adapter/base_http_adapter.ts

```diff
@@ -159,12 +193,88 @@ export abstract class BaseHttpAdapter implements Connection {
         removeRequestListeners()
       }
 
-      const config = this.config
-      function onSocket(socket: net.Socket): void {
-        // Force KeepAlive usage (workaround due to Node.js bug)
-        // https://github.com/nodejs/node/issues/47137#issuecomment-1477075229
-        socket.setKeepAlive(true, 1000)
-        socket.setTimeout(config.request_timeout, onTimeout)
+      function pipeStream(): void {
+        // if request.end() was called due to no data to send
+        if (request.writableEnded) {
+          return
+        }
+
+        const bodyStream = isStream(params.body)
+          ? params.body
+          : Stream.Readable.from([params.body])
+
+        const callback = (err: NodeJS.ErrnoException | null): void => {
+          if (err) {
+            removeRequestListeners()
+            reject(err)
+          }
+        }
+
+        if (params.compress_request) {
+          Stream.pipeline(bodyStream, Zlib.createGzip(), request, callback)
+        } else {
+          Stream.pipeline(bodyStream, request, callback)
+        }
+      }
+
+      const onSocket = (socket: net.Socket) => {
+        if (this.retry_expired_sockets) {
+          // if socket is reused
+          const socketInfo = this.known_sockets.get(socket)
+          if (socketInfo !== undefined) {
+            this.logger.trace({
+              module: 'Connection',
+              message: `Reused socket ${socketInfo.id}`,
+            })
+            // if a socket was reused at an unfortunate time,
+            // and is likely about to expire
+            const isPossiblyExpired =
+              Date.now() - socketInfo.last_used_time >
+              this.config.keep_alive.socket_ttl
+            if (isPossiblyExpired) {
+              this.logger.trace({
+                module: 'Connection',
+                message: 'Socket should be expired - terminate it',
+              })
+              this.known_sockets.delete(socket)
+              socket.destroy() // immediately terminate the connection
+              request.destroy()
+              reject(new Error(expiredSocketMessage))
+            } else {
+              this.logger.trace({
+                module: 'Connection',
+                message: `Socket ${socketInfo.id} is safe to be reused`,
+              })
+              this.known_sockets.set(socket, {
+                ...socketInfo,
+                last_used_time: Date.now(),
+              })
+              pipeStream()
+            }
+          } else {
+            const socketId = uuid.v4()
+            this.logger.trace({
+              module: 'Connection',
+              message: `Using a new socket ${socketId}`,
+            })
+            Object.assign(socket, { id: socketId })
```

Comment:
Could you answer the proposal here https://github.com/ClickHouse/clickhouse-js/pull/169#discussion_r1245571173 ?

Reply:
Thanks for pointing this out, this is rudimentary (we use WeakMap now).